File: 933.number-of-recent-calls.java
/*
 * @lc app=leetcode id=933 lang=java
 *
 * [933] Number of Recent Calls
 */

// @lc code=start

import java.util.ArrayDeque;
import java.util.Queue;

class RecentCounter {

    private Queue<Integer> queue;
    private static final int DURATION = 3000;

    public RecentCounter() {
        this.queue = new ArrayDeque<>();
    }
    
    public int ping(int t) {
        this.queue.add(t);
        int limit = t - DURATION;
        while (!this.queue.isEmpty() && this.queue.peek() < limit) {
            this.queue.remove();
        }
        return this.queue.size();
    }
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter obj = new RecentCounter();
 * int param_1 = obj.ping(t);
 */
// @lc code=end

File: 933.number-of-recent-calls.cpp
/*
 * @lc app=leetcode id=933 lang=cpp
 *
 * [933] Number of Recent Calls
 */

// @lc code=start

#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

class RecentCounter {
public:
    RecentCounter() { }
    
    int ping(int t) {
        timestamps.push(t);
        while (!timestamps.empty() && timestamps.front() < t - 3000) {
            timestamps.pop();
        }
        return timestamps.size();
    }

private:
    queue<int> timestamps;
};
/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter* obj = new RecentCounter();
 * int param_1 = obj->ping(t);
 */
// @lc code=end
